tests/end_to_end/test_account.py

```diff
@@ -0,0 +1,77 @@
+import pytest
+import pytest_asyncio
+from starknet_py.net.full_node_client import FullNodeClient
+
+TOTAL_SUPPLY = 10000 * 10**18
+TEST_AMOUNT = 10 * 10**18
+
+
+@pytest.fixture(scope="session")
+async def class_hashes():
+    """
+    All declared class hashes.
+    """
+    from kakarot_scripts.utils.starknet import get_declarations
+
+    return get_declarations()
+
+
+@pytest_asyncio.fixture(scope="module")
+async def erc20_token(deploy_solidity_contract, owner):
+    return await deploy_solidity_contract(
+        "UniswapV2",
+        "ERC20",
+        TOTAL_SUPPLY,
+        caller_eoa=owner.starknet_contract,
+    )
+
+
+@pytest.mark.asyncio(scope="session")
+@pytest.mark.AccountContract
+class TestAccount:
+    class TestAutoUpgradeOnTransaction:
+        async def test_should_upgrade_outdated_account_on_transfer(
+            self,
+            starknet: FullNodeClient,
+            invoke,
+            erc20_token,
+            owner,
+            other,
+            class_hashes,
+        ):
+            # Given an initial class
+            prev_class = await starknet.get_class_hash_at(
+                owner.starknet_contract.address
+            )
+            assert prev_class == class_hashes["account_contract"]
```

Comment:
what if `class_hashes["account_contract"]` and `class_hashes["account_contract_fixture"]` are the same? reading the test there is no guarantee that the class is well change.

That's why you need to do in this order

- `assert prev_class != target_class`
- kakarot.upgrade(target_class)
- `assert new_class == target_class`

so that it's sure that the class has been changed to match the target one stored in kakarot